src/DotNetWorker/Hosting/FunctionsApplicationBuilder.cs

```diff
@@ -0,0 +1,134 @@
+﻿using System;
+using System.Collections.Generic;
+using Microsoft.Azure.Functions.Worker;
+using Microsoft.Azure.Functions.Worker.Hosting;
+using Microsoft.Azure.Functions.Worker.Middleware;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Diagnostics.Metrics;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Extensions.Hosting;
+
+/// <summary>
+/// Placeholder
+/// </summary>
+public class FunctionsApplicationBuilder : IHostApplicationBuilder, IFunctionsWorkerApplicationBuilder
+{
+    private readonly HostApplicationBuilder _hostApplicationBuilder;
+    private readonly BootstrapHostBuilder _bootstrapHostBuilder;
+    private readonly IFunctionsWorkerApplicationBuilder _functionsWorkerApplicationBuilder;
+
+    private bool _hasBuilt = false;
+
+    internal const string SkipDefaultWorkerMiddlewareKey = "__SkipDefaultWorkerMiddleware";
+
+    internal FunctionsApplicationBuilder(string[]? args)
+    {
+        var configuration = new ConfigurationManager();
+
+        _hostApplicationBuilder = new HostApplicationBuilder(new HostApplicationBuilderSettings
+        {
+            Configuration = configuration
+        });
+
+        _bootstrapHostBuilder = new BootstrapHostBuilder(_hostApplicationBuilder);
+
+        // We'll set the default middleware up just before we build.
+        _bootstrapHostBuilder.Context.Properties[SkipDefaultWorkerMiddlewareKey] = true;
+
+        _bootstrapHostBuilder
+                   .ConfigureDefaults(args)
+                   .ConfigureFunctionsWorkerDefaults();
+
+        _functionsWorkerApplicationBuilder = InitializeHosting(_bootstrapHostBuilder);
+    }
+
+    /// <inheritdoc/>
+    public IDictionary<object, object> Properties => ((IHostApplicationBuilder)_hostApplicationBuilder).Properties;
+
+    IConfigurationManager IHostApplicationBuilder.Configuration => Configuration;
+
+    /// <inheritdoc/>
+    public ConfigurationManager Configuration => _hostApplicationBuilder.Configuration;
+
+    /// <inheritdoc/>
+    public IHostEnvironment Environment { get; private set; }
+
+    /// <inheritdoc/>
+    public ILoggingBuilder Logging => _hostApplicationBuilder.Logging;
+
+    /// <inheritdoc/>
+    public IMetricsBuilder Metrics => _hostApplicationBuilder.Metrics;
+
+    /// <inheritdoc />
+    public IServiceCollection Services => _hostApplicationBuilder.Services;
+
+    /// <summary>
+    /// Only available to AspNetCore extension (and maybe not even there in the future)
+    /// </summary>
+    internal IHostBuilder HostBuilder => _bootstrapHostBuilder;
```

Comment:
Doing this for now so that only AspNetCore extension has access to this. We could do something similar to what AspNet does with the `ConfigureHostBuilder` they expose which imparts some limitations. We'd need to do something similar to ensure this is never built and some things are not touched, etc. But for now, this gets us unblocked.

For reference: https://github.com/dotnet/aspnetcore/blob/559ada655240c2e6a0023ced978b291b05673b2c/src/DefaultBuilder/src/ConfigureHostBuilder.cs